291 Allow the use of formatters without the full UI context in React / React Native

mimi-acn opened the pull request:
# what was done so far
- added dedicated formatting context provider & hook
- non-breaking migration of current UI context to use this new provider for its formatters & added deprecation note
- added basic documentation examples
- adjust react example app
- adjust any current examples of UI provider using formatters
- add more documentation on how to use formatters in React regardless of ui component library


# what still needs to be done
- indicate which formatters are problematic in React Native

# issue: 
https://github.com/bonfhir/bonfhir/issues/291

mimi-acn committed:
feat: add formatters context in r4b

mimi-acn committed:
feat: add formatters context in r5

mimi-acn committed:
fix: missing s to formatters & add demo documentation

doc: starting doc on how to integrate in your UI

mimi-acn committed:
chore: add console.warn for use of formatters in UI context

mimi-acn committed:
chore: comment-out the logging for the time being

mimi-acn committed:
fix: adding removed hook on ui context

mimi-acn committed:
chore: move the formatters to the new hook in the basic react components

opiation commented:
This will eventually be blocked by #293 and I will address that issue so as to allow PRs like this one to eventually get merged.

mimi-acn committed:
chore: update template apps with Formatters provider

mimi-acn committed:
fix: replace never with unknown for better type continuity

mimi-acn committed:
fix: add formatter types suggestions when using format methods

mimi-acn committed:
doc: adjust base layout to use the new provider

mimi-acn committed:
doc: add changeset

opiation commented on docs/website/packages/react/components/use-fhir-formatters-context.md:
```diff
@@ -0,0 +1,282 @@
+
+  const handleOnLocaleSwitch = (changedTo: "en" | "es" | "fr") => {
+    setLocale(changedTo);
+  };
```
Nitpick: I think inlining `setLocale` reads better in docs.

e.g.:

```tsx
<button onClick={() => setLocale("en")}>EN</button>
```

It's almost imperative English, _On click, set the locale to "en"_. It also makes for shorter documentation :shrug:.

mimi-acn replied:
but I like to teach proper react naming conventions 😂

opiation replied:
I don't think I'd support any argument about this being a difference in naming convention apart from explicitly mentioning the _handling_ of the user action event. In addition though, I think this is mostly a difference in prose and I'm not willing to fight over style like this unless the code actually like something truly misleading. Your call.

opiation commented on packages/react/src/r5/context.ts:
```diff
@@ -21,25 +24,12 @@ export interface FhirUIContext {
-
-  return context;
-};
+export const useFhirUIContext = () => useContext(FhirUIContext);
```
Why exactly do we remove the null check assertion that was there before?

mimi-acn replied:
ah sorry, will put it back!

opiation commented on packages/core/src/r5/formatters.ts:
```diff
@@ -242,13 +243,18 @@ export class Formatter {
     const valueFormatter = this._formatters.get(type);
     if (!valueFormatter) {
-      throw new Error(`No formatter found for type '${type}'`);
+      if (value) {
```
Is it intentional that we also throw for _nullish_ values (e.g.: `""`, `false`)? If not, then we should explicitly check for `null` and `undefined`.

mimi-acn replied:
will add explicit checks

opiation commented on packages/react/src/r5/formatters/fhir-formatters-context.tsx:
```diff
@@ -0,0 +1,22 @@
+type FhirFormatters = Formatter; // redirecting in case we add strictly react-functionalities
+
+export const FhirFormattersContext = createContext<FhirFormatters>(
+  {} as FhirFormatters,
```
This would be a good place for a sane default

mimi-acn replied:
hmmm might only be confusing tough? Without a provider, this context is pretty much dead weight?

mimi-acn replied:
I tried to add them, hopefully it doesn't break stuff ahahaha

mimi-acn committed:
fix: add default for formatters context

fix: add null check in context hook
fix: move to prop the instance of the Formatter class

mimi-acn committed:
fix: formatter deconstruction

fix: no more need for never type

mimi-acn committed:
docs: shorthand version for documentation

opiation commented on packages/react/src/r4b/context.ts:
```diff
@@ -21,9 +24,7 @@ export interface FhirUIContext {
-export const FhirUIContext = createContext<FhirUIContext | undefined>(
-  undefined,
-);
+export const FhirUIContext = createContext<FhirUIContext>({} as FhirUIContext);
```
The this context's initial value should change to some default UI context value that does actually conform to the `FhirUIContext` interface or its type should be widen to `FhirUIContext | undefined` with its initial value set to `undefined`.

With this change, the error a developer will get when forgetting to provide a context value above with state that some property is missing because the empty object will circumvent the null check from `useFhirUIContext`  without providing the runtime guarantee that the context value actually conforms to `FhirUIContext`.

mimi-acn replied:
the problem with expecting types TContext | undefined is trailing the undefined everywhere you use the context

there is a "magic trick" I can do, so I'll submit that one, but you might not like it 😂

opiation commented on packages/react/src/r4b/provider.tsx:
```diff
@@ -1,33 +1,26 @@
-  }, [props.formatter]);
+// DO NOT EXPORT, this is to support the independent formatters context
+const FhirUIProviderWrapper: React.FC<DeprecationWrapperProps> = (props) => {
+  const { formatter } = useFhirFormatters();
```
Nitpick: Can we use a single noun everywhere around the formatter? I understand the utility composes a collection of type-specific formatters but the thing actually being used is a single instance of `Formatter`. I think we can also return just the formatter in that case and not an object with a `formatter` property. Unless I'm missing something, this could be:

```jsx
function Example() {
  const formatter = useFhirFormatter();

  return <>{formatter.format("HumanName", ...)}</>
}
```

mimi-acn replied:
done! had to do some juggling with names since FhirFormatter is a react component

opiation commented on packages/react/src/r4b/provider.tsx:
```diff
@@ -1,33 +1,26 @@
     <FhirUIContext.Provider
       value={{
-        formatter,
+        formatter: formatter,
```
Nitpick: This can use shorthand notation.

mimi-acn replied:
done!

mimi-acn committed:
fix: remove plural notation for formatter context

opiation committed:
feat: :sparkles: support parent-provided FHIR formatter

The goal here is to leverage any ancestor-provided FHIR formatter in
`FhirUIProvider` if one is available and none is passed in as a prop.
